### PR TITLE
Create new template - MemberEx

### PR DESCRIPTION
**Pull request by evanbtlr**
- Created new template;
- fixed small bugs;
- added tags for project search.

### CODE REVIEW

File: CSharp/VirtoCommerce.MemberExModule.Template/Data/Models/SupplierDataEntity.cs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        /// This method used to convert data type instance to domain model
+        /// </summary>
+        public override Member ToModel(Member member)
+        {
+            // Here you can write code for custom mapping
+            // supplier properties will be mapped in base method implementation by using value injection
+            var retVal = base.ToModel(member) as Supplier;
+
+            if (retVal != null)
+                retVal.Reviews = Reviews.OrderBy(x => x.Id).Select(x => x.ToModel(new SupplierReview())).ToList();
```
Comment (eMazeika) [inline CSharp/VirtoCommerce.MemberExModule.Template/Data/Models/SupplierDataEntity.cs]: use abstract type factory everywhere

File: CSharp/VirtoCommerce.MemberExModule.Template/Data/Repositories/SupplierRepository.cs [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+        public IQueryable<SupplierReviewDataEntity> SupplierReviews
+        {
+            get { return GetAsQueryable<SupplierReviewDataEntity>(); }
+        }
+
+        public override MemberDataEntity[] GetMembersByIds(string[] ids, string responseGroup = null, string[] memberTypes = null)
+        {
+            var retVal = base.GetMembersByIds(ids, responseGroup, memberTypes);
+            var reviews = SupplierReviews.Where(x => ids.Contains(x.SupplierId)).ToArray();
```
Comment (eMazeika) [inline CSharp/VirtoCommerce.MemberExModule.Template/Data/Repositories/SupplierRepository.cs]: The preferred syntax:
_ = SupplierReviews.Where(x => ids.Contains(x.SupplierId)).ToArray();

File: CSharp/VirtoCommerce.MemberExModule.Template/Data/Services/SupplierMemberService.cs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    /// </summary>
+    public class SupplierMemberService : CommerceMembersServiceImpl
+    {
+        public SupplierMemberService(Func<ICustomerRepository> repositoryFactory, IDynamicPropertyService dynamicPropertyService, ISecurityService securityService, IEventPublisher eventPublisher, ICommerceService commerceService)
+            : base(repositoryFactory, dynamicPropertyService, commerceService, securityService, eventPublisher)
+        {
+        }
+
+
+        //Override this method you can use for search members you custom tables and columns
```
Comment (eMazeika) [inline CSharp/VirtoCommerce.MemberExModule.Template/Data/Services/SupplierMemberService.cs]: //By overriding this method you can search by your custom tables and columns

File: CSharp/VirtoCommerce.MemberExModule.Template/Properties/AssemblyInfo.cs [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+// General Information about an assembly is controlled through the following 
+// set of attributes. Change these attribute values to modify the information
+// associated with an assembly.
+[assembly: AssemblyTitle("VirtoCommerce.MemberExModule.Template")]
+[assembly: AssemblyDescription("")]
+[assembly: AssemblyConfiguration("")]
+[assembly: AssemblyCompany("Бутлер И.С")]
```
Comment (eMazeika) [inline CSharp/VirtoCommerce.MemberExModule.Template/Properties/AssemblyInfo.cs]: ?

File: CSharp/VirtoCommerce.MemberExModule.Template/Properties/AssemblyInfo.cs [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+[assembly: AssemblyProduct("VirtoCommerce.MemberExModule.Template")]
+[assembly: AssemblyCopyright("Copyright © Бутлер И.С 2019")]
```
Comment (eMazeika) [inline CSharp/VirtoCommerce.MemberExModule.Template/Properties/AssemblyInfo.cs]: sorry, no

File: CSharp/VirtoCommerce.MemberExModule.Template/Web/Module.cs [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+using VirtoCommerce.Platform.Core.Common;
+using VirtoCommerce.Platform.Core.Modularity;
+using VirtoCommerce.Platform.Data.Infrastructure;
+using VirtoCommerce.Platform.Data.Infrastructure.Interceptors;
+
+namespace $safeprojectname$
+{
+    public class Module : ModuleBase
+    {
+        private readonly string _connectionString = ConfigurationHelper.GetConnectionStringValue("VirtoCommerce");
```
Comment (eMazeika) [inline CSharp/VirtoCommerce.MemberExModule.Template/Web/Module.cs:20]: add custom connection string, as always

File: CSharp/VirtoCommerce.MemberExModule.Template/Web/Scripts/blades/supplier-detail.html [hunk 7]
```diff
@@ -0,0 +1,18 @@
+<div class="blade-static __bottom" ng-include="'$(Platform)/Scripts/common/templates/create.tpl.html'" ng-if="blade.isNew"></div>
+<div class="blade-content">
+    <div class="blade-inner">
+        <div class="inner-block clearfix" ng-controller="$ext_safeprojectnamecamel$.supplierDetailController">
+            <form class="form" name="detailForm">
+                <div ng-init="setForm(detailForm)">
+                    <label class="form-label">Supplier Name</label>
```
Comment (eMazeika) [inline CSharp/VirtoCommerce.MemberExModule.Template/Web/Scripts/blades/supplier-detail.html:7]: localize

File: CSharp/VirtoCommerce.MemberExModule.Template/Web/Scripts/module.js [hunk 8]
```diff
@@ -0,0 +1,40 @@
+//Call this to register our module to main application
+var moduleName = "$ext_safeprojectnamecamel$";
+
+if (AppDependencies !== undefined) {
+    AppDependencies.push(moduleName);
+}
+
+angular.module(moduleName, [])
```
Comment (eMazeika) [inline CSharp/VirtoCommerce.MemberExModule.Template/Web/Scripts/module.js:8]: add dependency on customer module

File: CSharp/VirtoCommerce.MemberExModule.Template/Web/Scripts/module.js [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+    .run(['$rootScope', 'virtoCommerce.customerModule.memberTypesResolverService',
+        function ($rootScope, memberTypesResolverService) {
+
+            // add JobTitle field to Contact detail blade
+            var contactInfo = memberTypesResolverService.resolve("Contact");
+            contactInfo.detailBlade.metaFields.unshift({
+                name: 'jobTitle',
+                title: "JobTitle",
+                valueType: "ShortText"
+            });
```
Comment (eMazeika) [inline CSharp/VirtoCommerce.MemberExModule.Template/Web/Scripts/module.js:18]: what's that?

File: CSharp/VirtoCommerce.MemberExModule.Template/Web/module.manifest [hunk 10]
```diff
@@ -0,0 +1,33 @@
+﻿<?xml version="1.0" encoding="utf-8" ?>
+<module>
+    <id>MemberExtensionSampleModule</id>
```
Comment (eMazeika) [inline CSharp/VirtoCommerce.MemberExModule.Template/Web/module.manifest:3]: replace

File: CSharp/VirtoCommerce.MemberExModule.Template/Web/module.manifest [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+    <version>1.0.0</version>
+    <platformVersion>2.13.44</platformVersion>
+    <dependencies>
+        <dependency id="VirtoCommerce.Core" version="2.25.21" />
+        <dependency id="VirtoCommerce.Customer" version="2.14.21" />
+    </dependencies>
+
+    <title>MemberExtensionSampleModule</title>
```
Comment (eMazeika) [inline CSharp/VirtoCommerce.MemberExModule.Template/Web/module.manifest:11]: replace

File: CSharp/VirtoCommerce.MemberExModule.Template/Web/module.manifest [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+    <description>Enter your module description here.</description>
+    <authors>
+        <author>evanb</author>
```
Comment (eMazeika) [inline CSharp/VirtoCommerce.MemberExModule.Template/Web/module.manifest:14]: replace

File: CSharp/VirtoCommerce.MemberExModule.Template/Web/module.manifest [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+    </authors>
+    <owners>
+        <owner></owner>
+    </owners>
+
+    <releaseNotes>First version.</releaseNotes>
+    <copyright>Copyright © 2019 . All rights reserved</copyright>
```
Comment (eMazeika) [inline CSharp/VirtoCommerce.MemberExModule.Template/Web/module.manifest:21]: add placeholders

File: CSharp/VirtoCommerce.MemberExModule.Template/Web/module.manifest [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+    <tags>extension module</tags>
+  
+    <assemblyFile>MemberExtensionSampleModule.Web.dll</assemblyFile>
+    <moduleType>MemberExtensionSampleModule.Web.Module, MemberExtensionSampleModule.Web</moduleType>
```
Comment (eMazeika) [inline CSharp/VirtoCommerce.MemberExModule.Template/Web/module.manifest:25]: .

File: CSharp/VirtoCommerce.Module.Template/Web/Scripts/module.js [hunk 15]
```diff
@@ -1,4 +1,4 @@
-﻿//Call this to register our module to main application
+//Call this to register our module to main application
```
Comment (eMazeika) [inline CSharp/VirtoCommerce.Module.Template/Web/Scripts/module.js]: //Call this to register your module to main application

File: CSharp/VirtoCommerce.OrderExModule.Template/Web/Content/help/index.html [hunk 16]
```diff
@@ -50,12 +50,7 @@ <h2>Problems and fixs</h2>
...
                 <p>You must recreate the migration or add a new one in order to get the current state of the models. It is also necessary to remove changes to entities that have already been made. Only what we did in the extension should remain.</p>
                 <p>For migration, use the following commands, indicating the type of migration configuration, for multiple migrations in one project: </p>
-                <pre class="indent">Add-Migration {MigrateName} 
-    -ConfigurationTypeName {MigrateConfiguration} 
-    -ConnectionString "Data Source=(local);Initial Catalog=VirtoCommerce2_Samples;
-        Persist Security Info=True;User ID=virto;Password=virto;
-        MultipleActiveResultSets=True;Connect Timeout=420" 
-    -ConnectionProviderName "System.Data.SqlClient"</pre>
+                <pre class="indent">Add-Migration {MigrateName} -ConfigurationTypeName {MigrateConfiguration} -ConnectionString "Data Source=(local);Initial Catalog=VirtoCommerce2_Samples;Persist Security Info=True;User ID=virto;Password=virto;MultipleActiveResultSets=True;Connect Timeout=420" -ConnectionProviderName "System.Data.SqlClient"</pre>
```
Comment (eMazeika) [inline CSharp/VirtoCommerce.OrderExModule.Template/Web/Content/help/index.html]: change VirtoCommerce2_Samples to VirtoCommerce2

File: CSharp/VirtoCommerce.PriceExModule.Template/Web/Content/help/index.html [hunk 17]
```diff
@@ -48,13 +48,8 @@ <h2>Problems and fixs</h2>
...
                 <p>For migration, use the following commands, indicating the type of migration configuration, for multiple migrations in one project: </p>
-                <pre class="indent">Add-Migration {MigrateName} 
-    -ConfigurationTypeName {MigrateConfiguration} 
-    -ConnectionString "Data Source=(local);Initial Catalog=VirtoCommerce2_Samples;
-        Persist Security Info=True;User ID=virto;Password=virto;
-        MultipleActiveResultSets=True;Connect Timeout=420" 
-    -ConnectionProviderName "System.Data.SqlClient"</pre>
-                <p>Where {MigrateName} is the name of the migration, in the current case it is Initial. And the configuration {MigrateConfiguration} is equal to $safeprojectname$.Migrations.Cart.Configuration or $safeprojectname$.Migrations.Order.Configuration.</p>
+                <pre class="indent">Add-Migration {MigrateName} -ConnectionString "Data Source=(local);Initial Catalog=VirtoCommerce2_Samples;Persist Security Info=True;User ID=virto;Password=virto;MultipleActiveResultSets=True;Connect Timeout=420" -ConnectionProviderName "System.Data.SqlClient"</pre>
```
Comment (eMazeika) [inline CSharp/VirtoCommerce.PriceExModule.Template/Web/Content/help/index.html]: same change

File: CSharp/VirtoCommerce.MemberExModule.Template/VirtoCommerce.MemberExModule.Template.vstemplate [hunk 18]
```diff
@@ -0,0 +1,44 @@
+<?xml version="1.0" encoding="utf-8"?>
+<VSTemplate Version="3.0.0" Type="ProjectGroup" xmlns="http://schemas.microsoft.com/developer/vstemplate/2005" xmlns:sdk="http://schemas.microsoft.com/developer/vstemplate-sdkextension/2010">
+  <TemplateData>
+    <Name>Virto Commerce 2.x Member Module extension</Name>
```
Comment (eMazeika) [inline CSharp/VirtoCommerce.MemberExModule.Template/VirtoCommerce.MemberExModule.Template.vstemplate]: Virto Commerce 2.x Customer Module extension

File: CSharp/VirtoCommerce.MemberExModule.Template/VirtoCommerce.MemberExModule.Template.vstemplate [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+    <Description>A project for extending Virto Commerce member models</Description>
```
Comment (eMazeika) [inline CSharp/VirtoCommerce.MemberExModule.Template/VirtoCommerce.MemberExModule.Template.vstemplate]: A project for extending VirtoCommerce.Customer module with new (Supplier) member type

**Review by eMazeika — CHANGES_REQUESTED** [review]
Any UI for SupplierReview?
widgets for standard member fields missing: Addresses, Phones, Emails, Notes, Groups, DynamicProperties.

**Review by eMazeika — APPROVED** [review]

File: CSharp/VirtoCommerce.MemberExModule.Template/Data/Models/SupplierDataEntity.cs (hunk 1 above)
Comment (evanbtlr) [reply CSharp/VirtoCommerce.MemberExModule.Template/Data/Models/SupplierDataEntity.cs]: fixed

File: CSharp/VirtoCommerce.MemberExModule.Template/Data/Repositories/SupplierRepository.cs (hunk 2 above)
Comment (evanbtlr) [reply CSharp/VirtoCommerce.MemberExModule.Template/Data/Repositories/SupplierRepository.cs]: fixed

File: CSharp/VirtoCommerce.MemberExModule.Template/Data/Services/SupplierMemberService.cs (hunk 3 above)
Comment (evanbtlr) [reply CSharp/VirtoCommerce.MemberExModule.Template/Data/Services/SupplierMemberService.cs]: fixed

File: CSharp/VirtoCommerce.MemberExModule.Template/Properties/AssemblyInfo.cs (hunk 4 above)
Comment (evanbtlr) [reply CSharp/VirtoCommerce.MemberExModule.Template/Properties/AssemblyInfo.cs]: Company fixed

File: CSharp/VirtoCommerce.MemberExModule.Template/Properties/AssemblyInfo.cs (hunk 5 above)
Comment (evanbtlr) [reply CSharp/VirtoCommerce.MemberExModule.Template/Properties/AssemblyInfo.cs]: fixed

File: CSharp/VirtoCommerce.Module.Template/Web/Scripts/module.js (hunk 15 above)
Comment (evanbtlr) [reply CSharp/VirtoCommerce.Module.Template/Web/Scripts/module.js]: fixed

File: CSharp/VirtoCommerce.OrderExModule.Template/Web/Content/help/index.html (hunk 16 above)
Comment (evanbtlr) [reply CSharp/VirtoCommerce.OrderExModule.Template/Web/Content/help/index.html]: fixed - in all projects

File: CSharp/VirtoCommerce.PriceExModule.Template/Web/Content/help/index.html (hunk 17 above)
Comment (evanbtlr) [reply CSharp/VirtoCommerce.PriceExModule.Template/Web/Content/help/index.html]: fixed

File: CSharp/VirtoCommerce.MemberExModule.Template/VirtoCommerce.MemberExModule.Template.vstemplate (hunk 18 above)
Comment (evanbtlr) [reply CSharp/VirtoCommerce.MemberExModule.Template/VirtoCommerce.MemberExModule.Template.vstemplate]: fixed

File: CSharp/VirtoCommerce.MemberExModule.Template/VirtoCommerce.MemberExModule.Template.vstemplate (hunk 19 above)
Comment (evanbtlr) [reply CSharp/VirtoCommerce.MemberExModule.Template/VirtoCommerce.MemberExModule.Template.vstemplate]: Fixed

**Pull request merged**